### ISSUE TITLE
Multi-Frame DICOM load performace decreased substantially in 4.13.2+

### ISSUE THREAD

**Issue by timbra339** [InsightSoftwareConsortium/ITK#1542]
I've updated my ITK libraries from 4.13.0 to 5.0.1 recently and noticed a quite substantial drop in DICOM data load performance.

I've looked a bit into this using the attached python test-file which loads a ~800MB multi-frame DICOM file. Here are the resulting load times using various versions of ITK (used python-wheels for that):

ITK Version: 4.13.0  (python-package itk==4.13.1.post1)
Loading took **2.3250107765197754** secs.

ITK Version: 4.13.2
Loading took **16.011502027511597** secs.

ITK Version: 5.0.0
Loading took 15.990726470947266 secs.

ITK Version: 5.0.1
Loading took 15.931232690811157 secs.

ITK Version: 5.1.0
Loading took 15.805387496948242 secs.

It looks like the performance drop cause was introduced with ITK 4.13.2, whose release notes state several changes to GDCM, so that appears to be a likely cause? The issues seems to be still in the current codebase, maybe it can be fixed for the upcoming 5.1 version? I can supply the test data if needed.

```python
# -*- coding: utf-8 -*-

import itk
import time

print(f"ITK Version: {itk.Version.GetITKVersion()}")

reader = itk.ImageFileReader[itk.Image[itk.US, 3]].New(FileName = r"Volume_0000.dcm")
start = time.time()
reader.Update()
stop = time.time()

print(f"Loading took {stop-start} secs.")
```

**Comment by dzenanz** [comment]
Please give a link to the test data.

**Comment by timbra339** [comment]
Uploading the test data. 
I've zeroed out the actual image content to make the file compressible, the runtimes for 4.13.0 and 5.1 are still  as reported.
[Volume_0000.zip](https://github.com/InsightSoftwareConsortium/ITK/files/4035688/Volume_0000.zip)

**Comment by thewtex** [comment]
ITK 5.1 RC 1 contains many GDCM updates.

You can install the wheels for testing with:

```
pip install --upgrade --pre itk
```

@malaterre

**Comment by timbra339** [comment]
@thewtex: This command installs the version 5.1rc1.post1 on my machine, which was the one that produced the 

ITK Version: 5.1.0
Loading took 15.805387496948242 secs.

output. Should there be a more recent wheel instead?

**Comment by dzenanz** [comment]
@issakomi might also pitch in.

**Comment by issakomi** [comment]
@dzenanz  thank you for CC, my app may affected too
i confirm, tested with own C++ small prog, the issue doesn't  exist in 4.13.1 and exists in 4.13.2.
itkGDCMImageIO didn't change, but GDCM was updated. The slowdown happens in
_image.GetBuffer( (char*)pointer )_ in GDCMImageIO::Read()

**Comment by issakomi** [comment]
Seems (not 100% sure now) that this change in gdcmBitmap.h (in _bool Bitmap::TryRAWCodec(char *buffer, bool &lossyflag) const_ for particular image, the same in all codecs)
```
-codec.SetNeedOverlayCleanup( AreOverlaysInPixelData() );
+codec.SetNeedOverlayCleanup( AreOverlaysInPixelData() || UnusedBitsPresentInPixelData() );
```

makes the difference, not yet 100% sure, will check more later.

P.S. BTW, the test image Volume_0000 is strange, it has negative spacing, but it doesn't matter in particular case, tested with other images.

**Comment by thewtex** [comment]
> Should there be a more recent wheel instead?

Yes, *5.1rc1.post1* is the current latest wheel.

**Comment by issakomi** [comment]
Rather sure now, the last post above was correct, image reader spend most time with cleaning unused bits... I have also this change in my production clone, were interesting to know if it is required, reading is really faster without it.
(time in ms)
```
$ ./test Volume_0000.dcm 
InternalReadImageInformation=165
DoOverlayCleanup TRUE
DoOverlayCleanup dt=10055
ITK  reader dt=11545
```

**Comment by malaterre** [comment]
> ```
> -codec.SetNeedOverlayCleanup( AreOverlaysInPixelData() );
> +codec.SetNeedOverlayCleanup( AreOverlaysInPixelData() || UnusedBitsPresentInPixelData() );
> ```
> 
> makes the difference, not yet 100% sure, will check more later.

The above change is actually required to be 100% conformant. There is not possible revert on my side for this piece of code. I can be more verbose here if needed.

Two solutions comes to mind:
1. (hackish) Introduced a new option `SkipCleanupUnusedBits` in gdcm.ImageReader API
2. Long term solution, check why bit masking implementation is so slow. I am pretty sure there are room for improvement here.

**Comment by timbra339** [comment]
> P.S. BTW, the test image Volume_0000 is strange, it has negative spacing, but it doesn't matter in particular case, tested with other images.

Ah, thanks for pointing that out, this was a bug in the dicom exporter code.

**Comment by issakomi** [comment]
> The above change is actually required to be 100% conformant.

I understand,

[PS3.5 8.8.1](http://dicom.nema.org/medical/dicom/current/output/chtml/part05/chapter_8.html#sect_8.1.1)
`Receiving applications may not assume anything about the contents of unused bits, and in particular may not assume that they are zero, or that they contain sign extension bits.`

I've already done an on/off UI option in my app, still in doubt about the default value.

**Comment by issakomi** [comment]
During making the option to skip cleaning of unused bits, i mentioned an issue in GDCM. There are already several options, e.g. _ForceRescaleInterceptSlope_. IMHO, they are declared twice, once [here, in gdcmImageHelper.h](https://github.com/malaterre/GDCM/blob/25d540987d9a7eede51236e5fd45b06f9c20ac5e/Source/MediaStorageAndFileFormat/gdcmImageHelper.h#L156) and [here, in gdcmImageHelper.cpp](https://github.com/malaterre/GDCM/blob/25d540987d9a7eede51236e5fd45b06f9c20ac5e/Source/MediaStorageAndFileFormat/gdcmImageHelper.cxx#L48) again, aren't they?

**Comment by issakomi** [comment]
Deleted previous post, sorry, was wrong, overseen one thing.
It were probably also good to inform Slicer team about performance drop. BTW, i've preferred to set the variable to skip cleaning by default in my app.

**Comment by lassoan** [comment]
We experienced severe slowdown in 3D Slicer, too (#1586). The timing and the amount of slowdown is so significant that it makes users very unhappy.

Clearing unused bits is probably the right thing to do. Long time ago "unused" bits were actually used to store overlay bits but this option was retired in 2004 (http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.9.2.html), so nowadays there should be no useful information in those bits.

The problem is that clearing of unused bits is done very inefficiently. To be exact, these two lines (and especially writing) takes most of the time when you load an image:

https://github.com/malaterre/GDCM/blob/25d540987d9a7eede51236e5fd45b06f9c20ac5e/Source/MediaStorageAndFileFormat/gdcmImageCodec.cxx#L544-L546

Maybe processing the stream in blocks of tens of kilobytes instead of 2 bytes would solve the problem.

**Comment by thewtex** [comment]
Addressed in #1608
